metadata-builders/src/main/java/com/linkedin/metadata/builders/search/MLModelIndexBuilder.java

```diff
@@ -0,0 +1,203 @@
+    }
+
+    @Nonnull
+    private List<MLModelDocument> getDocumentsToUpdateFromSnapshotType(@Nonnull MLModelSnapshot mlModelSnapshot) {
+        final MLModelUrn urn = mlModelSnapshot.getUrn();
+        return mlModelSnapshot.getAspects().stream().map(aspect -> {
+            if (aspect.isCaveatsAndRecommendations()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getCaveatsAndRecommendations());
+            } else if (aspect.isEthicalConsiderations()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getEthicalConsiderations());
+            } else if (aspect.isEvaluationData()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getEvaluationData());
+            } else if (aspect.isInstitutionalMemory()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getInstitutionalMemory());
+            } else if (aspect.isIntendedUse()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getIntendedUse());
+            } else if (aspect.isMetrics()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getMetrics());
+            } else if (aspect.isMLModelFactorPrompts()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getMLModelFactorPrompts());
+            } else if (aspect.isMLModelProperties()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getMLModelProperties());
+            } else if (aspect.isOwnership()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getOwnership());
+            } else if (aspect.isQuantitativeAnalyses()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getQuantitativeAnalyses());
+            } else if (aspect.isStatus()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getStatus());
+            } else if (aspect.isTrainingData()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getTrainingData());
+            }
+            return null;
+        }).filter(Objects::nonNull).collect(Collectors.toList());
+    }
+
+    @Nonnull
+    private MLModelDocument getDocumentToUpdateFromAspect(MLModelUrn urn, CaveatsAndRecommendations caveatsAndRecommendations) {
+        return setUrnDerivedFields(urn);
+    }
+
+    @Nonnull
+    private MLModelDocument getDocumentToUpdateFromAspect(MLModelUrn urn, EthicalConsiderations ethicalConsiderations) {
+        return setUrnDerivedFields(urn);
+    }
+
+    @Nonnull
+    private MLModelDocument getDocumentToUpdateFromAspect(MLModelUrn urn, EvaluationData evaluationData) {
+        final MLModelDocument doc = setUrnDerivedFields(urn);
+
+        if (evaluationData.hasEvaluationData()) {
+            DatasetUrnArray datasetUrns = evaluationData.getEvaluationData()
+                .stream()
+                .map(BaseData::getDataset)
+                .collect(Collectors.toCollection(DatasetUrnArray::new));
+            doc.setEvaluationDatasets(datasetUrns);
```

Comment:
Not an issue but would love to hear your thoughts. Storing the dataset urn itself could limit search capability. One needs to know the entire dataset URN to be able to find an ML model that is evaluated on it. Is that desired? One way to expand the search capability is to allow search over dataset name. For that we could index just the dataset name (obtained from the urn). Alternative approach is to keep the index mapping as it is but add a subfield under `evaluationDatasets` that stores the names (we will need an appropriate analyzer to extract the name from dataset URN). Similar comment for `trainingDatasets`.
Let me know what you think.

Reply:
I took the same approach here as the upstreams array in Datasets. I definitely think it makes sense to support searching by the name as opposed to the whole URN. I think most users won't have the whole URN available when searching. I think I like the analyzer approach more here since it allows for other parts of the URN to be indexed as well if wanted rather than restricting the search capabilities to only search by name. 

What are your thoughts on keeping the analyzer for the name as a keyword vs extending ngram support for the name search? Most of the other name type fields seem to support ngram for search.

Reply:
Not sure what you mean by "keeping the analyzer for the name as a keyword". My original suggestion was to store urns in field `evaluationDatasets` and have a subfield under this field with an analyzer that will store the relevant `name` obtained from the urn. something like
`"evaluationDatasets": {
  "type": "keyword",
  "fields": {
    "name": {
      "type": "text",
      "analyzer": "extract_name_from_dataseturn"
    }
  },
  "normalizer": "my_normalizer"
}`
Here `extract_name_from_dataseturn` is the analyzer which will extract name from dataset urn.
You can add more subfields if you wish to provide ngram support for the name search.
Again, happy to hear your thoughts. We can keep it simple in this PR by only storing dataset urns in `evaluationDatasets` and add subfields later. Also let me know if you need help with defining `extract_name_from_dataseturn` analyzer.

Reply:
Okay, that answers my question :) I was asking if the subfield should look like what you put + the ngram support or if it should be like:

"fields": { "name" { "type": "keyword"...}

Reply:
If the subfield stores exactly what the (parent) field (here presumably `urn`) represents then you could use the `keyword` type for the subfield. In this case, if I am not wrong, subfield `name` is supposed to store different token(s) hence we need an analyzer.

Reply:
Yeah we need an analyzer either way, type: keyword operates differently from type: text so I was just asking if name should keep keyword type or be more flexible with text & ngram support.

Reply:
Sounds good. Will you be able to push the corresponding change in this PR? If you would like we can defer it to a later PR.

Reply:
Yeah I can push out the commit on this PR.

Reply:
I implemented this using a stop filter & pattern tokenizer to split out the actual urn components without the "urn", "li", and entity type being searchable.